internal/graph/check.go

```diff
@@ -945,6 +953,499 @@ func trySendUsersetsAndDeleteFromMap(ctx context.Context, usersetsMap usersetsMa
+	ds storage.RelationshipTupleReader,
+	req *ResolveCheckRequest,
+	allowedUserTypeRestrictions []*openfgav1.RelationReference,
+	usersetMessageChan chan usersetMessage) {
+	ctx, span := tracer.Start(ctx, "nestedUsersetFirstLevelLookupObject")
+	defer span.End()
+
+	storeID := req.GetStoreID()
+	reqTupleKey := req.GetTupleKey()
+
+	opts := storage.ReadUsersetTuplesOptions{
+		Consistency: storage.ConsistencyOptions{
+			Preference: req.GetConsistency(),
+		},
+	}
+
+	iter, err := ds.ReadUsersetTuples(ctx, storeID, storage.ReadUsersetTuplesFilter{
+		Object:                      reqTupleKey.GetObject(),
+		Relation:                    reqTupleKey.GetRelation(),
+		AllowedUserTypeRestrictions: allowedUserTypeRestrictions,
+	}, opts)
+	if err != nil {
+		span.RecordError(err)
+		concurrency.TrySendThroughChannel(ctx, usersetMessage{
+			userset: "",
+			err:     err,
+		}, usersetMessageChan)
+		return
+	}
+
+	filteredIter := storage.NewConditionsFilteredTupleKeyIterator(
+		storage.NewFilteredTupleKeyIterator(
+			storage.NewTupleKeyIteratorFromTupleIterator(iter),
+			validation.FilterInvalidTuples(typesys),
+		),
+		checkutil.BuildTupleKeyConditionFilter(ctx, req.GetContext(), typesys),
+	)
+	defer filteredIter.Stop()
+
+	for {
+		t, err := filteredIter.Next(ctx)
+		if err != nil {
+			if errors.Is(err, storage.ErrIteratorDone) || errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				break
+			}
+
+			// error encountered.  No need to process further
+			span.RecordError(err)
+			concurrency.TrySendThroughChannel(ctx, usersetMessage{
+				userset: "",
+				err:     err,
+			}, usersetMessageChan)
+			return
+		}
+		usersetName, relation := tuple.SplitObjectRelation(t.GetUser())
+		if relation == "" {
+			// This should never happen because the filter should only allow userset with relation.
+			err = fmt.Errorf("unexpected userset %s with no relation", t.GetUser())
+			span.RecordError(err)
+			concurrency.TrySendThroughChannel(ctx, usersetMessage{
+				userset: "",
+				err:     err,
+			}, usersetMessageChan)
+			return
+		}
+		concurrency.TrySendThroughChannel(ctx, usersetMessage{
+			userset: usersetName,
+			err:     nil,
+		}, usersetMessageChan)
+	}
+}
+
+func processUsersetMessage(message usersetMessage,
+	primarySortedSet storage.SortedSet,
+	secondarySortedSet storage.SortedSet) (bool, error) {
+	if message.err != nil {
+		return false, message.err
+	}
+	primarySortedSet.Add(message.userset)
+	if secondarySortedSet.Exists(message.userset) {
+		return true, nil
+	}
+	return false, nil
+}
+
+func processNestedUsersetFirstLevelMessage(ctx context.Context,
+	req *ResolveCheckRequest,
+	userToUsersetMessageChan,
+	objectToUsersetMessageChan chan usersetMessage) (*ResolveCheckResponse, storage.SortedSet, storage.SortedSet, error) {
+	usersetFromUser := storage.NewSortedSet()
+	usersetFromObject := storage.NewSortedSet()
+
+	userToUsersetDone := false
+	objectToUsersetDone := false
+
+	for {
```

Comment:
```suggestion
	for !userToUsersetDone && !objectToUsersetDone {
```

and then you don't need `if`s in line 1305 and 1335